1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py

```diff
@@ -0,0 +1,320 @@
+
+if (cwd / ".git").is_dir():
+    root_dir = cwd
+
+else:
+    root_dir = None
+    for parent in cwd.parents:
+        if (parent / ".git").is_dir():
+            root_dir = parent
+            break
+
+# Check if a Git root directory was found
+if root_dir is None:
+    raise FileNotFoundError("No Git root directory found.")
+
+
+# ## Custom Imports
+
+# In[ ]:
+
+
+sys.path.append(str((root_dir / "1.develop_vision_models").resolve(strict=True)))
+
+from ImageDataset import ImageDataset
+from models.fnet_nn_2d import Net
+from ModelTrainer import ModelTrainer
+from transforms.CropNPixels import CropNPixels
+from transforms.StandardScaler import StandardScaler
+
+
+# ## Set random seeds
+
+# In[1]:
+
+
+random.seed(0)
+np.random.seed(0)
+
+mlflow.log_param("random_seed", 0)
+
+
+# # Inputs
+
+# In[ ]:
+
+
+# Nuclei crops path of treated nuclei in the Dapi channel with all original pixel values
+treated_dapi_crops = root_dir / "vision_nuclear_speckle_prediction/treated_nuclei_dapi_crops"
+
+# Nuclei crops path of nuclei in the Gold channel with all original pixel values
+gold_crops = root_dir / "vision_nuclear_speckle_prediction/gold_cropped_nuclei"
+
+# Paths to original nuclear speckle data
+data_dir = root_dir / "nuclear_speckles_data"
+nuclear_mask_dir = (data_dir / "Nuclear_masks").resolve(strict=True)
+sc_profiles_path = list((data_dir / "Preprocessed_data/single_cell_profiles").resolve(strict=True).glob("*feature_selected*.parquet"))
+
+# Load single-cell profile data
+scdfs = [pd.read_parquet(sc_path) for sc_path in sc_profiles_path if sc_path.is_file()]
+scdfs = pd.concat(scdfs, axis=0).reset_index(drop=True)
+
+
+# # Outputs
+
+# In[ ]:
+
+
+figure_path = pathlib.Path("fnet_validation_images_temp")
+figure_path.mkdir(parents=True, exist_ok=True)
+
+metrics_path = pathlib.Path("metrics")
+metrics_path.mkdir(parents=True, exist_ok=True)
+
+model_path = pathlib.Path("model")
+model_path.mkdir(parents=True, exist_ok=True)
+
+
+# In[ ]:
+
+
+description = "Here we leverage the 2d fnet architecture in https://doi.org/10.1038/s41592-018-0111-2 to predict the GOLD stain from cropped DAPI Nuclei Images. We retain all pixel values in the cropped images"
+mlflow.set_tag("mlflow.note.content", description)
+
+
+# # Image Generation Functions
+
+# In[ ]:
+
+
+def format_img(_tensor_img):
+    """Reshapes an image and rescales pixel values from the StandardScaler transform."""
+
+    mean = trainer.val_dataset.dataset.input_transform[0].mean
+    std = trainer.val_dataset.dataset.input_transform[0].std
+
+    return (torch.squeeze(_tensor_img) * std + mean).to(torch.uint16).cpu().numpy()
+
+
+# In[ ]:
+
+
+def evaluate_and_format_imgs(_input, _target):
+
+    single_input = input.unsqueeze(1).to(device)
+    single_target = target.to(device)
+
+    model.eval()
+    with torch.no_grad():
+        # Forward Pass
+        output = model(single_input)
+
+    return format_img(single_input), format_img(single_target), format_img(output)
```

Comment:
Consider adding docstrings and type hints here

Reply:
Good idea, I added type hints here. I think together the code and the one function description summarize what is happening here. e.g. the model computes the output given the input `output = model(single_input)`